### PR TITLE
Adding inconsistentStructName change

### LINKED ISSUE #2: Add check for verify consistent instance name accross all the methods of the same struct
All the methods of the same struct should use the same variable name for the object instance, we have inconsistencies in the code about this and it should be consistent.

### PR DESCRIPTION
**Pull request by jespino**
Fixes #2

### CODE REVIEW

File: inconsistentStructName/inconsistentStructName.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package inconsistentStructName
```
Comment (agnivade) [inline inconsistentStructName/inconsistentStructName.go]: It's not really a struct name. The spec calls it a receiver: https://golang.org/ref/spec#Method_declarations. I would suggest `inconsistentReceiverName`.

File: inconsistentStructName/inconsistentStructName.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"go/ast"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name: "inconsistentStructName",
+	Doc:  "check for inconsistent struct name in the methods definition",
```
Comment (agnivade) [inline inconsistentStructName/inconsistentStructName.go]: check for inconsistent receiver names in the methods of a struct

File: inconsistentStructName/inconsistentStructName.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	Run:  run,
+}
+
+type firstVarName = struct {
+	Name string
+	Node ast.Node
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	firstVarNameForType := make(map[string]firstVarName)
```
Comment (agnivade) [inline inconsistentStructName/inconsistentStructName.go]: firstReceiverNameForType

File: inconsistentStructName/inconsistentStructName.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+					if !ok {
+						var starExpr *ast.StarExpr
+						starExpr, ok = field.Type.(*ast.StarExpr)
+						if ok {
+							typeIdent, ok = starExpr.X.(*ast.Ident)
+						}
+					}
+
+					if ok {
+						varName := field.Names[0].Name
```
Comment (agnivade) [inline inconsistentStructName/inconsistentStructName.go]: Check for len > 0

File: inconsistentStructName/inconsistentStructName.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+						if _, exists := firstVarNameForType[typeIdent.Name]; !exists {
```
Comment (agnivade) [inline inconsistentStructName/inconsistentStructName.go]: We are duplicating the map access here. I would suggest to do it like:
```go
mapEntry, ok := firstVarNameForType[typeIdent.Name]
if !ok {
	firstVarNameForType[typeIdent.Name] = firstVarName{
			Name: varName,
			Node: node,
		}
	return true
}
if mapEntry.Name != varName {
	// use mapEntry.Name
}
```

File: inconsistentStructName/inconsistentStructName.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+							firstVarNameForType[typeIdent.Name] = firstVarName{
+								Name: varName,
+								Node: node,
+							}
+						} else if firstVarNameForType[typeIdent.Name].Name != varName {
+							otherVarName := firstVarNameForType[typeIdent.Name].Name
+							otherVarPosition := pass.Fset.Position(firstVarNameForType[typeIdent.Name].Node.Pos())
+							pass.Reportf(node.Pos(), "Different variable name used for the struct %s in different methods, using %s here but was named as %s before at %s", typeIdent.Name, varName, otherVarName, otherVarPosition.String())
```
Comment (agnivade) [inline inconsistentStructName/inconsistentStructName.go]: Different receiver name .. : using %s here ...

File: inconsistentStructName/inconsistentStructName.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	firstVarNameForType := make(map[string]firstVarName)
+	for _, file := range pass.Files {
+		ast.Inspect(file, func(node ast.Node) bool {
+			switch funDecl := node.(type) {
+			case *ast.FuncDecl:
+				if funDecl.Recv != nil && len(funDecl.Recv.List) > 0 && len(funDecl.Recv.List[0].Names) > 0 {
+					field := funDecl.Recv.List[0]
+					typeIdent, ok := field.Type.(*ast.Ident)
```
Comment (agnivade) [inline inconsistentStructName/inconsistentStructName.go]: I would re-arrange the checks from StarExpr first. Then the check for *ast.Ident becomes easy.

```go
field := funDecl.Recv.List[0]
expr := field.Type
starExpr, ok := field.Type.(*ast.StarExpr)
if ok {
	expr = starExpr.X
}
typeIdent, ok := expr.(*ast.Ident)
if !ok {
	return false
}
// use typeIdent from here
```

**Review by agnivade — CHANGES_REQUESTED** [review]
I see that you have been busy in your vacation :)

File: inconsistentStructName/inconsistentStructName.go (hunk 4 above)
Comment (jespino) [reply inconsistentStructName/inconsistentStructName.go]: That is already checked in the first if

**Comment by jespino** [comment]
I think I addressed everything here.

**Review by agnivade — APPROVED** [review]
LGTM

**Pull request merged**